woocommerce/woocommerce
# In App: Allow installing free products from any In App Marketplace page

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

- Adds new `<InstallFlow>` component. This listens for changes in `query`, and shows the corresponding modal to the user.
- Allows merchants to install any free and .org hosted product from any page, while creating the order on the go. Currently this mocks order creation process.
- Adds new **Add to Store** button on search pages. You need to switch to the Woo.com branch as well.



### How to test the changes in this Pull Request:

1. Switch `MARKETPLACE_HOST` in `marketplace/components/constants.ts` to `https://woocommerce.test`
2. Build assets with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build`
3. Make sure you **don't** have WooPayments installed. We'll install it as part of our test, having it already installed causes errors.
4. Disconnect store from Woo.com by clicking **Avatar > Disconnect** from store
5. [On the search page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions), find a product with the **Add to Store** button and click it. It should show a warning for requiring a connection.
6. Complete the connection. Please note when you come back to your store, the modal should open automatically.
7. In the next modal, click Install. This will first ping Woo.com, create an order and then start downloading the product from WordPress.org repository. 
8. When the installation is done and we're allowed to activate (sometimes we're not allowed), you should see the **Activate** button. Clicking the button will activate the product and then show an option to see the plugin list, to set up the plugin. 
9. On Woo.com, delete your customer's billing address and try to install the plugin again. You won't see the **Add to Store** on the same product again, because it's now installed. 
10. Trying to install a new product, should first show a warning and then redirect user to Woo.com with the product added to cart.
11. When it's done, we should test installations with a product key. Find a product key from the API requests in [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) and then visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions&install=PRODUCT_KEY_STARTS_WITH_W00.
12. On the [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) click **Install** in the table and confirm it works.
